src/backend/vaapi.rs

```diff
@@ -224,6 +210,56 @@ mod surface_pool {
 
     use crate::Resolution;
 
+    /// A VA Surface obtained from a `[SurfacePool]`.
+    ///
+    /// The surface will automatically be returned to its pool upon dropping, provided the pool still
+    /// exists and the surface is still compatible with it.
+    pub struct PooledSurface<D: SurfaceMemoryDescriptor> {
+        surface: Option<Surface<D>>,
+        pool: Weak<RefCell<SurfacePool<D>>>,
+    }
+
+    impl<D: SurfaceMemoryDescriptor> PooledSurface<D> {
+        fn new(surface: Surface<D>, pool: &Rc<RefCell<SurfacePool<D>>>) -> Self {
+            Self {
+                surface: Some(surface),
+                pool: Rc::downgrade(pool),
+            }
+        }
+
+        /// Detach this surface from the pool. It will not be returned, and we can dispose of it
+        /// freely.
+        pub fn detach_from_pool(mut self) -> Surface<D> {
+            // `unwrap` will never fail as `surface` is `Some` up to this point.
+            self.surface.take().unwrap()
+        }
+    }
+
+    impl<D: SurfaceMemoryDescriptor> Borrow<Surface<D>> for PooledSurface<D> {
+        fn borrow(&self) -> &Surface<D> {
```

Comment:
I wonder why we have Borrow instead of AsRef?

Reply:
`Borrow<Self>` is implemented by `Self`, but `AsRef<Self>` is not. Using `Borrow` here allows to pass a `Surface` directly, like we used to do before, something that would not be possible with `AsRef` (we would have to encapsulate the `Surface` somewhere before).